File: LeetCode75_LevelTwo/Implementing_Queue_Using_Stacks/Implementing_Queue_Using_Stacks.cc
// 232. Implement Queue using Stacks C++
#include <vector>

class MyQueue {
    std::vector<int> v1;
    std::vector<int> v2;
public:
    MyQueue() {
        
    }
    
    void push(int x) {
        v1.emplace_back(x);
    }
    
    int pop() {
        peek();
        int output = v2.back();
        v2.pop_back();
        return output;
    }
    
    int peek() {
        if (v2.empty()) {
            while (!v1.empty()) {
                int tmp = v1.back();
                v1.pop_back();
                v2.emplace_back(tmp);
            }
        }
        return v2[v2.size() - 1];
    }
    
    bool empty() {
        return v1.empty() && v2.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */

File: LeetCode75_LevelTwo/Implementing_Queue_Using_Stacks/Implementing_Queue_Using_Stacks.py
# 232. Implement Queue using Stacks Python
class MyQueue(object):
    def __init__(self):
        self.stack = []
        self.temp = []
    def push(self, x):
        """
        :type x: int
        :rtype: None
        """
        self.stack.append(x)

    def pop(self):
        """
        :rtype: int
        """
        self.peek()
        return self.temp.pop()
        
    def peek(self):
        """
        :rtype: int
        """
        if not self.temp:
            while self.stack:
                self.temp.append(self.stack.pop())
        return self.temp[-1]

    def empty(self):
        """
        :rtype: bool
        """
        return not self.stack and not self.temp
        


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()
